src/modules/qt/qimage_wrapper.cpp

```diff
@@ -180,98 +228,103 @@ int refresh_qimage( producer_qimage self, mlt_frame frame )
+	}
 
-	// optimization for subsequent iterations on single picture
-	if ( image_idx != self->image_idx || width != self->current_width || height != self->current_height )
-		self->current_image = NULL;
+	int has_alpha = sourceImage.hasAlphaChannel();
+	if ( has_alpha )
+	{
+		format = mlt_image_rgb24a;
+	}
 
 	// If we have a qimage and need a new scaled image
-	if ( self->qimage && ( !self->current_image || ( format != mlt_image_none && format != mlt_image_glsl && format != self->format ) ) )
+	if ( !self->current_image || ( format != mlt_image_none && format != mlt_image_glsl && format != self->format ) )
 	{
 		QString interps = mlt_properties_get( properties, "rescale.interp" );
 		bool interp = ( interps != "nearest" ) && ( interps != "none" );
-		QImage *qimage = static_cast<QImage*>( self->qimage );
-		int has_alpha = qimage->hasAlphaChannel();
 		QImage::Format qimageFormat = has_alpha ? QImage::Format_ARGB32 : QImage::Format_RGB32;
 
 		// Note - the original qimage is already safe and ready for destruction
-		if ( qimage->format() != qimageFormat )
+		if ( sourceImage.format() != qimageFormat )
 		{
-			QImage temp = qimage->convertToFormat( qimageFormat );
-			qimage = new QImage( temp );
-			self->qimage = qimage;
+			sourceImage = sourceImage.convertToFormat( qimageFormat );
 			mlt_cache_item_close( self->qimage_cache );
-			mlt_service_cache_put( MLT_PRODUCER_SERVICE( producer ), "qimage.qimage", qimage, 0, ( mlt_destructor )qimage_delete );
+			mlt_service_cache_put( MLT_PRODUCER_SERVICE( producer ), "qimage.qimage", new QImage(sourceImage), 0, ( mlt_destructor )qimage_delete );
 			self->qimage_cache = mlt_service_cache_get( MLT_PRODUCER_SERVICE( producer ), "qimage.qimage" );
 		}
-		QImage scaled = interp? qimage->scaled( QSize( width, height ), Qt::IgnoreAspectRatio, Qt::SmoothTransformation ) :
-			qimage->scaled( QSize(width, height) );
 
-		// Convert scaled image to target format (it might be premultiplied after scaling).
-		scaled = scaled.convertToFormat( qimageFormat );
+		// Scale image to requested size
+		QImage scaled = interp? sourceImage.scaled( width, height, Qt::IgnoreAspectRatio, Qt::SmoothTransformation ) :
+			sourceImage.scaled( width, height );
 
 		// Store width and height
 		self->current_width = width;
 		self->current_height = height;
 
 		// Allocate/define image
+        int image_size;
+#if QT_VERSION >= 0x050200
+		if ( has_alpha )
+		{
+			image_size = 4 * width * height;
+			self->format = mlt_image_rgb24a;
+			scaled = scaled.convertToFormat( QImage::Format_RGBA8888 );
+		}
+		else
+		{
+			image_size = 3 * width * height;
+			self->format = mlt_image_rgb24;
+#if QT_VERSION >= 0x051300
+			scaled.convertTo( QImage::Format_RGB888 );
+#else
+			scaled = scaled.convertToFormat( QImage::Format_RGB888 );
+#endif
+		}
+		self->current_image = ( uint8_t * )mlt_pool_alloc( image_size );
+		memcpy( self->current_image, scaled.constBits(), image_size);
```

Comment:
you should at least check that scaled.sizeInBytes() is same or bigger than image_size so you don't read out of bounds.